src/pages/authpages/SignupPage.tsx

```diff
@@ -1,78 +1,297 @@
-import Input from '../../components/Input';
-import Button from '../../components/Button';
+import Input from '../../components/usercomponents/Input';
 import GoogleBtn from '../../components/authcomponents/GoogleBtn';
+import Button from '../../components/usercomponents/Button';
+import { ColorRing } from 'react-loader-spinner';
+import Modal from 'react-modal';
+import axios from 'axios';
+import {
+  setEmail,
+  setUsername,
+  setFirstname,
+  setLastname,
+  setPassword,
+  setLoading,
+  setErrorMessage,
+  setIsOpen,
+  setNavigation,
+  setEmailError,
+  setUsernameError,
+  setFirstnameError,
+  setLastnameError,
+  setPasswordError,
+} from '../../slices/signup.slice';
+import { useAppDispatch, useAppSelector } from '../../hooks/Hooks';
+import { useEffect } from 'react';
+import { Link, Navigate } from 'react-router-dom';
 
 const Signup = () => {
+  const dispatch = useAppDispatch();
+  const {
+    email,
+    username,
+    firstname,
+    lastname,
+    password,
+    isLoading,
+    errorMessage,
+    isOpen,
+    isNavigating,
+    emailError,
+    usernameError,
+    firstnameError,
+    lastnameError,
+    passwordError,
+  } = useAppSelector((state) => state.signup);
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    switch (name) {
+      case 'email':
+        dispatch(setEmail(value));
+        break;
+      case 'username':
+        dispatch(setUsername(value));
+        break;
+      case 'firstname':
+        dispatch(setFirstname(value));
+        break;
+      case 'lastname':
+        dispatch(setLastname(value));
+        break;
+      case 'password':
+        dispatch(setPassword(value));
+        break;
+      default:
+        break;
+    }
+  };
+
+  const validateForm = () => {
+    let valid = true;
+
+    if (
+      !email.match(/^([a-z\d.-]+)@([a-z\d-]+)\.([a-z]{2,3})(\.[a-z]{2,3})?$/)
+    ) {
+      dispatch(setEmailError('Invalid email format'));
+      valid = false;
+    } else if (email === '') {
+      dispatch(setEmailError('Email is required'));
+      valid = false;
+    } else {
+      dispatch(setEmailError(''));
+    }
+
+    if (username === '' || username.length < 5) {
+      dispatch(setUsernameError('Username must be at least 5 characters'));
+      valid = false;
+    } else {
+      dispatch(setUsernameError(''));
+    }
+
+    if (firstname === '') {
+      dispatch(setFirstnameError('Firstname is required'));
+      valid = false;
+    } else {
+      dispatch(setFirstnameError(''));
+    }
+
+    if (lastname === '') {
+      dispatch(setLastnameError('Lastname is required'));
+      valid = false;
+    } else {
+      dispatch(setLastnameError(''));
+    }
+
+    if (password === '' || password.length < 6) {
+      dispatch(setPasswordError('Password must be at least 6 characters'));
+      valid = false;
+    } else {
+      dispatch(setPasswordError(''));
+    }
+
+    return valid;
+  };
```

Comment:
Find a better way to validate input data. Hint: checkout `zod`. After that create a utils folder in the root > schemas folder then schema validators there